runtime/compact.c

```diff
@@ -35,95 +35,73 @@
-
-  /* Use Ecolor (q) == 0 instead of Is_block (q) because q could be an
-     inverted pointer for an infix header (with Ecolor == 2). */
-  if (Ecolor (q) == 0 && Is_in_heap (q)){
-    switch (Ecolor (Hd_val (q))){
-    case 0:
-    case 3: /* Pointer or header: insert in inverted list. */
-      *p = Hd_val (q);
-      Hd_val (q) = (header_t) p;
-      break;
-    case 1: /* Infix header: make inverted infix list. */
-      /* Double inversion: the last of the inverted infix list points to
-         the next infix header in this block.  The last of the last list
-         contains the original block header. */
-      {
-        /* This block as a value. */
-        value val = (value) q - Infix_offset_val (q);
-        /* Get the block header. */
-        word *hp = (word *) Hp_val (val);
-
-        while (Ecolor (*hp) == 0) hp = (word *) *hp;
-        CAMLassert (Ecolor (*hp) == 3);
-        if (Tag_ehd (*hp) == Closure_tag){
-          /* This is the first infix found in this block. */
-          /* Save original header. */
-          *p = *hp;
-          /* Link inverted infix list. */
-          Hd_val (q) = (header_t) ((word) p | 2);
-          /* Change block header's tag to Infix_tag, and change its size
-             to point to the infix list. */
-          *hp = Make_ehd (Wosize_bhsize (q - val), Infix_tag, 3, (uintnat) 0);
+  header_t h;
+
+  CAMLassert (((uintnat) p & 3) == 0);
+
+  if (Is_block (q)){
+    h = Hd_val (q);
+    switch (Color_hd (h)){
+    case Caml_white:
+      if (Tag_hd (h) == Infix_tag){
+        value realvalue = (value) q - Infix_offset_val (q);
+        if (Is_black_val (realvalue)){
+          break;
         }else{
-          CAMLassert (Tag_ehd (*hp) == Infix_tag);
-          /* Point the last of this infix list to the current first infix
-             list of the block. */
-          *p = (word) &Field (val, Wosize_ehd (*hp)) | 1;
-          /* Point the head of this infix list to the above. */
-          Hd_val (q) = (header_t) ((word) p | 2);
-          /* Change block header's size to point to this infix list. */
-          *hp = Make_ehd (Wosize_bhsize (q - val), Infix_tag, 3, (uintnat) 0);
+          CAMLassert (Is_in_heap (realvalue));
         }
+      }else{
+        CAMLassert (Is_in_heap (q));
       }
+      /* FALL THROUGH */
+    case Caml_gray:
+      CAMLassert (Is_in_heap (q));
+      /* [q] points to some inverted list, insert it. */
+      *p = h;
+      Hd_val (q) = eptr (p);
       break;
-    case 2: /* Inverted infix list: insert. */
-      *p = Hd_val (q);
-      Hd_val (q) = (header_t) ((word) p | 2);
+    case Caml_black:
+      /* [q] points to an out-of-heap value. Leave it alone. */
+      break;
+    case Caml_blue:
+      /* We found a pointer to a free block. This cannot happen. */
+      CAMLassert (0);
       break;
     }
   }
 }
 
 void caml_invert_root (value v, value *p)
 {
+  CAMLassert (Is_long (*p) || Is_in_heap (*p) || Is_black_val (*p));
```

Comment:
This assertion is wrong in naked pointers mode.
```suggestion
#ifdef NO_NAKED_POINTERS
  CAMLassert (Is_long (*p) || Is_in_heap (*p) || Is_black_val (*p));
#endif
```

Reply:
That said, note that this assertion is failing in CI currently. This indicates that even in no-naked-pointers mode, some root contains a naked pointer (or is inverted several times?), or something else fishy.

Reply:
Actua
```suggestion
  CAMLassert (Is_long (*p) || Is_in_heap (*p) || Is_black_val (*p) ||
              Tag_val(*p) == Infix_tag);
```

Actually, these assertion failures are simply related to white infix headers (see #9735).

Reply:
I dislike those complicated assertions.  You're  trying to check that `*p` is an OK OCaml value, but if it isn't the GC would have bombed already.

Reply:
These complicated assertions are usually the by-product of debugging sessions.